### PR TITLE
Check that url endswith .git rather than containing it anywhere in the

### PR DESCRIPTION
**Pull request by Crazybus**
Fixes: #55

### CODE REVIEW

**Comment by Crazybus** [comment]
I wasn't able to find a way to test this easily since the integration tests all require real repos to test with. Ideally we could create a new test repo with ".git" in the repository name.

**Comment by guyzmo** [comment]
Actually, it's not needed for the bugfix you've done (and it's necessary only once when working on the services implementation).

I really got to write some documentation on how to implement the tests. So let's explain it a bit here, so I'll have something stashed for when I'll write the doc ☺

Basically, git-repo has two layers:

* `git_repo.repo`: is where all the command line arguments are parsed, the service instanciated and is user of the service API
* `git_repo.services`: defines the services
 * `git_repo.services.service`: which contains: 
  * the base class for each specialized service,  
  * common methods that are service agnostic,
  * contains the entry point for the service selection mechanism.
 * `git_repo.services.ext.*`: which contains each specialized service (and registers it)

So the tests are actually matching that distinction:

* [`tests.integration.test_main`](https://github.com/guyzmo/git-repo/blob/master/tests/integration/test_main.py): 

it tests `git_repo.repo` stuff. What that mean is that it mocks up the input argument dictionary (that you can see when you call `git-repo -vvvvv`), using [a helper method](https://github.com/guyzmo/git-repo/blob/master/tests/helpers.py#L264,L270) run the call against it and exposes what the `git_repo.services.service` based API would receive as arguments, and then specializes the test using a [test case in `tests.integration.test_main`](https://github.com/guyzmo/git-repo/blob/master/tests/integration/test_main.py#L86-L88).

As a code is worth 1000 words, the test for your change could be done by calling `clone`:

``` python
    def test_clone(self):
        repo_slug_branch, seen_args = self.main_clone('guyzmo/test.git.test', 0)
        assert ('guyzmo', 'test.git.test', 'master') == repo_slug_branch
```

So as you can see, we're only testing the first layer, and thus there is no need for the repository to exist. There are some actions that run the `git` command though to create a temporary git repository to test things like guessing names based on the repository.

* [`tests.integration.test_(github|gitlab|bitbucket)`](https://github.com/guyzmo/git-repo/blob/master/tests/integration/test_github.py)

There the tests only starts at the services API (so we don't care about what's being done at the `git_repo.repo` level), and it needs to work against the real API.

But hopefully, you need to setup the *real* conditions for the test only once. Thanks to [betamax](https://github.com/sigmavirus24/betamax), when the `tests/integration/cassettes/*.json` file matching the test is missing, the test case will run the real HTTP interaction and record them as a JSON file that gets stored there. If the json file exists, then it will play the recorded version and no HTTP request will be sent by the test runner.

**Comment by Crazybus** [comment]
Thanks for the in-depth explanation! Unfortunately this is exactly what I tried first and it passed the test just fine. The problem is that the `_guess_repo_slug` is only called when it is unable to identify the repo properly. When you set the repo name like that (what I tried too) it is able to detect it just fine and doesn't need to "guess" what the repo name is by parsing the remote urls. 
```
        if not lookup_repository:
            service = RepositoryService.get_service(None, self.target)
            service.connect()
        else:
            # Try to resolve existing repository path
            try:
                try:
                    repository = Repo(os.path.join(self.path, self.repo_name or ''))
                except NoSuchPathError:
                    repository = Repo(self.path)
            except InvalidGitRepositoryError:
                raise FileNotFoundError('Cannot find path to the repository.')
            service = RepositoryService.get_service(repository, self.target)
            if not self.repo_name:
                self._guess_repo_slug(repository, service)
```
I tried creating a new repo and doing the cloning however betamax wasn't recording anything and the tests were always passing again. Any hints?

I was however able to reproduce it in one of the tests using real git commands which I have updated the PR with. 

```
git-repo/tests/integration/test_main.py:333: assert ('guyzmo', '.git-repo') == ('guyzmo', '.git-')
```

File: tests/integration/test_main.py [hunk 1]
```diff
@@ -325,6 +325,17 @@ def test_request_list__no_repo_slug__https(self, capsys, caplog):
...
+    def test_request_list__no_repo_slug__https_dot_git(self, capsys, caplog):
+        from subprocess import call
+        call(['git', 'init', '-q', self.tempdir.name])
+        call(['git', '--git-dir={}/.git'.format(self.tempdir.name), 'remote', 'add', 'github', 'https://github.com/guyzmo/.git-repo'])
+        repo_slug, seen_args = self.main_request_list(rc=0, args={})
+        assert ('guyzmo', '.git-repo') == repo_slug
+        assert dict() == seen_args
+        out, err = capsys.readouterr()
+        assert out ==  '  1\tdesc1                                                       \thttp://request/1\n  2\tdesc2                                                       \thttp://request/2\n  3\tdesc3                                                       \thttp://request/3\n'
+        assert 'id' in caplog.text and 'title' in caplog.text and 'URL' in caplog.text
```
Comment (guyzmo) [inline tests/integration/test_main.py]: The three lines above (335-337) are not really useful to validate the test, as it's covered by the surrounding tests.
What's important really is the assert on line 334.

So I'd be for removing those ☺

File: tests/integration/test_main.py [hunk 2]
```diff
@@ -325,6 +325,17 @@ def test_request_list__no_repo_slug__https(self, capsys, caplog):
         assert out ==  '  1\tdesc1                                                       \thttp://request/1\n  2\tdesc2                                                       \thttp://request/2\n  3\tdesc3                                                       \thttp://request/3\n'
         assert 'id' in caplog.text and 'title' in caplog.text and 'URL' in caplog.text
 
+    def test_request_list__no_repo_slug__https_dot_git(self, capsys, caplog):
```
Comment (guyzmo) [inline tests/integration/test_main.py]: for the name, I'd be happy if you made a reference to the issue something like:

``` python
    def test_request_list__no_repo_slug__https_dot_git_fix__issue55(self)
```

**Comment by guyzmo** [comment]
About your concerns:

> The problem is that the _guess_repo_slug is only called when it is unable to identify the repo properly

indeed, and that's actually the intended behaviour, `_guess_repo_slug()` is only called when there's no slug passed as argument.
The `<repo>/<slug>` CLI argument always has precedence.

> When you set the repo name like that (what I tried too) it is able to detect it just fine and doesn't need to "guess" what the repo name is by parsing the remote urls

Yup, actually what you've done in your test is the right thing! I'm sorry I forgot about that when I told you about the general case, but indeed it's a case where you need to create a real git repository to check that the guessing works fine.

> I tried creating a new repo and doing the cloning however betamax wasn't recording anything and the tests were always passing again. Any hints?

Well for your use case you don't have to run a betamax recording, as I said, the `test_main` stuff does not care about the network side, you just check what's going on between the CLI argument dict and the `git_repo.services.service` implementation.

**Comment by guyzmo** [comment]
not sure why it's not appearing here, but it's merged

**Pull request closed without merging**